enclave-interface/README.md

```diff
@@ -0,0 +1,369 @@
+    Bls461,
+    /// Barreto-Lynn-Scott embedding degree 24 with equivalent security of AES-192
+    Bls24,
+    /// Barreto-Lynn-Scott embedding degree 48 with equivalent security of AES-256 with 581 bit prime
+    /// see https://tools.ietf.org/id/draft-kato-threat-pairing-01.html#rfc.section.3
+    Bls48,
+    /// Barreto Naehrig with prime field 254 bits. About 100 bits of security
+    Bn254,
+    /// Barreto Naehrig with prime field 256 bits. See https://cryptojedi.org/papers/dclxvi-20100714.pdf
+    /// and https://moderncrypto.org/mail-archive/curves/2016/000740.html for more details
+    BnP256,
+    /// Barreto Naehrig with prime field 384 bits. See https://cryptojedi.org/papers/dclxvi-20100714.pdf
+    BnP384,
+    /// Barreto Naehrig with prime field 512 bits as specified in ISO15946-5
+    BnP512,
+    /// Barreto Naehrig with prime field 638 bits as specified in
+    /// https://fidoalliance.org/specs/fido-v2.0-rd-20180702/fido-ecdaa-algorithm-v2.0-rd-20180702.html#supported-curves-for-ecdaa
+    BnP638,
+    /// NIST P-224 (secp224r1)
+    EcP224,
+    /// NIST P-256 (secp256r1, prime256v1)
+    EcP256,
+    /// NIST P-384 (secp384r1)
+    EcP384,
+    /// NIST P-521 (secp521r1)
+    EcP521,
+    /// secp256k1
+    EcK256,
+    /// Brainpool 224
+    EcBP224,
+    /// BrainPool 256
+    EcBP256,
+    /// BrainPool 384
+    EcBP384,
+    /// BrainPool 512
+    EcBP512,
+}
+```
+
+Some capabilities offer multiple options like `DeriveDiffieHellman` which can be either PKCS#3 DHParameter structure
+or Elliptic-Curve based.
+
+This model allows the flexibility to add new operations and types without changing the APIs.
+
+Each operation will return an Enclave Message that contains the results or an error upon failure. For example, 
+Encryption will return the ciphertext and possibly an authentication tag. Signing returns the signature.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+For developers intimately familiar with enclave programming, this API may be a more expensive/complicated scheme than
+using the methods they are already used to using. Care must be taken so this scheme does not change frequency with each
+enclave provider that is to be supported.
+
+Another possibility is that this approach is too flexible and requires intimate knowledge about crypto algorithms.
+To mitigate this, predefined ciphers can be created for end consumers like RSA-3072-PSS-SHA256 or
+AES-256-GCM or AES-128-CBC-HMAC-SHA256 or XCHACHA20-POLY1305. This reduces algorithmic agility that is an inherent problem
+with many cryptographic libraries. 
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+# Prior art
+[prior-art]: #prior-art
+This provider model is extremely common in the crypto world with implementations like [PKCS11](http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/os/pkcs11-base-v2.40-os.html)
+Microsoft CNG, Java JCA, [Parsec](https://github.com/parallaxsecond/parsec), and others. 
```

Comment:
May want to consider linking webkms, https://github.com/w3c-ccg/webkms/issues/4

Reply:
Possibly. Thanks for linking us from them. I hope they have something more concrete soon. Looking at it now, it doesn’t have much. Maybe we should invite them to this discussion?!?

Reply:
Here's something a bit more concrete... we're still finalizing the API and didn't want to write the spec bits until we ship something more stable to customers.

These docs aren't great, but they're the best we have unless someone wants to start diving into the code:

https://github.com/digitalbazaar/webkms-client/#api-reference